backend/service/topology/cache.go

```diff
@@ -0,0 +1,49 @@
+package topology
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// Perfroms leader election by acquiring a postgres advisory lock.
+// Once the lock is acquired topology caching is started.
+func (c *client) acquireTopologyCacheLock() {
+	// We create our own connection to use for acquiring the advisory lock
+	// If the connection is severed for any reason the advisory lock will automatically unlock
+	conn, err := c.db.Conn(context.Background())
+	if err != nil {
+		c.log.Fatal("Unable to connect to the database", zap.Error(err))
+	}
+
+	// Infinitely try to acquire the advisory lock
+	// Once the lock is acquired we start caching, this is a blocking operation
+	for {
+		c.log.Debug("trying to acquire advisory lock")
+
+		// Advisory locks only take a arbitrary bigint value
+		// In this case 100 is the lock for topology caching
+		if tryAdvisoryLock(conn, 100) {
+			c.log.Debug("acquired the advisory lock, starting to cache topology now...")
```

Comment:
another thing, it would be better if each individual topology updater (e.g. AWS, K8s, etc) had its own lock to distribute the work among different instances. good TODO.

Reply:
Interesting thought, that would make the locking service more compelling also. Ill leave some comments.